src/xmipp/libraries/py_xmipp/deepDenoising/augmentators.py

```diff
@@ -0,0 +1,138 @@
+    if bool(random.getrandbits(1)):
+      batchX[i] = np.flipud(batchX[i])
+      batchY[i] = np.flipud(batchY[i])
+  return batchX, batchY
+
+def _random_90degrees_rotation( batchX, batchY, rotations=[0, 1, 2, 3]):
+  for i in range(batchX.shape[0]):
+    num_rotations = random.choice(rotations)
+    batchX[i] = np.rot90(batchX[i], num_rotations)
+    batchY[i] = np.rot90(batchY[i], num_rotations)
+  return batchX, batchY
+
+def _random_rotation( batchX, batchY, max_angle=25.):
+  for i in range(batchX.shape[0]):
+    if bool(random.getrandbits(1)):
+      # Random angle
+      angle = random.uniform(-max_angle, max_angle)
+      batchX[i] = scipy.ndimage.interpolation.rotate(batchX[i], angle,reshape=False, mode="reflect")
+      batchY[i] = scipy.ndimage.interpolation.rotate(batchY[i], angle,reshape=False, mode="reflect")
+  return batchX, batchY
+
+def _random_blur( batchX, batchY, sigma_max):
+  for i in range(batchX.shape[0]):
+    if bool(random.getrandbits(1)):
+      # Random sigma
+      sigma = random.uniform(0., sigma_max)
+      batchX[i] = scipy.ndimage.filters.gaussian_filter(batchX[i], sigma)
+      batchY[i] = scipy.ndimage.filters.gaussian_filter(batchY[i], sigma)
+  return batchX, batchY
+
+
+def _mismatch_projection( batchX, batchY, p=0.05):
+  maxMismatch= 1+int(batchX.shape[1]*.01)
+  for i in range(batchX.shape[0]):
+    if np.random.rand()<p:
+      angle = random.uniform(-maxMismatch,-maxMismatch)
+      batchX[i] = scipy.ndimage.interpolation.rotate(batchX[i], angle,reshape=False, mode="reflect")
+  return batchX, batchY
+  
+
+def initPointsRadius(img_shape, aboveInsteadBelow=True, radiusFraction=0.9):
+
+  xv, yv,= np.meshgrid(np.arange( img_shape[0] ), np.arange( img_shape[1] ), indexing="xy")
+  allCoords= np.stack([xv, yv], axis=-1).reshape((-1,2))
+  central_coords= np.array([img_shape[0]//2, img_shape[1]//2] )
+  distMat= np.squeeze( np.sqrt( np.sum((central_coords- allCoords)**2, axis=1))  )
+  distThr= int(radiusFraction*(img_shape[0]/2))
+  if aboveInsteadBelow:
+    return allCoords[ distMat>distThr ].astype( np.int64)
+  else:
+    return allCoords[ distMat<distThr ].astype( np.int64)
+
+  
+def generateReverseNormalizationFunction(img_shape, radiusFraction=0.8):
+  pointsBelow= initPointsRadius(img_shape, aboveInsteadBelow=False, radiusFraction=radiusFraction)
+  def reverseNormalization(imgs, refs):
+    imgsNorm= np.zeros_like(imgs)
+    for i in range(imgs.shape[0]):
+      img, ref= imgs[i], refs[i]
+      innerParRef= ref[ pointsBelow[...,0], pointsBelow[...,1] ]
+      innerParImg= img[ pointsBelow[...,0], pointsBelow[...,1] ]
+      innerMeanRef, innerStdRef= np.mean(innerParRef), np.std(innerParRef)
+      innerMeanImg, innerStdImg= np.mean(innerParImg), np.std(innerParImg)
+      imgsNorm[i]= (img- (innerMeanImg-innerMeanRef))/(innerStdImg/innerStdRef)
+    return imgsNorm
+  return reverseNormalization
+    
+def generateEmptyParticlesFunction(img_shape, prob=0.2):
```

Comment:
prob is unused

Reply:
it is actually used in 
def swapParticlesForEmptyParticles1
if np.random.rand()<prob:

which is defined inside the function